bugsnag-android-core/src/main/java/com/bugsnag/android/SynchronizedStreamableStore.kt

```diff
@@ -0,0 +1,44 @@
+package com.bugsnag.android
+
+import android.util.JsonReader
+import java.io.File
+import java.io.InputStreamReader
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.withLock
+
+internal class SynchronizedStreamableStore<T : JsonStream.Streamable>(
+    private val file: File
+) {
+
+    private val lock = ReentrantReadWriteLock()
+
+    fun persist(streamable: T, logger: Logger): Boolean {
+        lock.writeLock().withLock {
+            return try {
+                file.writer().use {
+                    streamable.toStream(JsonStream(it))
+                    true
+                }
+            } catch (exc: Throwable) {
+                logger.w("Failed to persist user information", exc)
+                false
+            }
+        }
+    }
+
+    fun load(logger: Logger): T? {
+        lock.readLock().withLock {
+            return try {
+                val user = file.reader().use {
+                    User.fromReader(JsonReader(it))
+                }
+                @Suppress("UNCHECKED_CAST")
+                return user as T
+            } catch (exc: Throwable) {
```

Comment:
This seems too generic for a catch statement considering we control the implementations - it seems reasonable for the eventual caller to catch an `IOException` and handle it, rather than handling a null response.

Same for the `persist` catch - I think this can be passed up for the caller to handle.

Reply:
Updated to throw the exception and changed test assertions accordingly.